Get exposure time and analogue gain values when using Auto mode

Hello,

I’m currently trying to get the effective values of exposure time and analogue gain when their control mode is set to `Auto`.

- Is it possible to access the current exposure time and analogue gain values while `Auto` mode?
- Alternatively, after switching from `Auto` to `Manual`, is there a way to get the last values that were automatically selected?

I couldn’t find a way to access this information. Any guidance would be appreciated.

Thank you

Thank you,
For now, with the command `ros2 run camera_ros camera_node --ros-args --log-level camera:=debug -p camera:=0 -p ExposureTimeMode:=0`

I get the error : 
```
camera_node: /usr/local/include/libcamera/libcamera/controls.h:204: T libcamera::ControlValue::get() const [with T = libcamera::Span<const long int, 18446744073709551615>; typename std::enable_if<(libcamera::details::is_span<U>::value || std::is_same<std::basic_string_view<char>, typename std::remove_cv< <template-parameter-1-1> >::type>::value), std::nullptr_t>::type <anonymous> = nullptr]: Assertion `isArray_' failed.
```

the complete logs are : 
```
      [0:55:12.036551098] [58]  INFO Camera camera_manager.cpp:340 libcamera v0.6.0+rpt20251202                                                                                                                                                                                                   
[0:55:12.037739857] [72]  INFO RPI pisp.cpp:720 libpisp version v1.3.0 9ba67e6680f0 17-03-2026 (10:49:38)                                                                                                                                                                                   
[0:55:12.041311635] [72]  INFO IPAProxy ipa_proxy.cpp:180 Using tuning file /usr/local/share/libcamera/ipa/rpi/pisp/imx500.json                                                                                                                                                             
[0:55:12.048716654] [72]  INFO Camera camera_manager.cpp:223 Adding camera '/base/axi/pcie@1000120000/rp1/i2c@88000/imx500@1a' for pipeline handler rpi/pisp                                                                                                                                
[0:55:12.048817228] [72]  INFO RPI pisp.cpp:1181 Registered camera /base/axi/pcie@1000120000/rp1/i2c@88000/imx500@1a to CFE device /dev/media1 and ISP device /dev/media4 using PiSP variant BCM2712_D0                                                                                     
[DEBUG] [1773768403.203641973] [camera]: found camera by id: 0                                                                                                                                                                                                                              
[DEBUG] [1773768403.204289602] [camera]: default Viewfinder stream configuration: "800x600-XRGB8888/sYCC"                                                                                                                                                                                   
[DEBUG] [1773768403.204572991] [camera]:                                                                                                                                                                                                                                                    
>> stream formats:                                                                                                                                                                                                                                                                          
   - R8 (32x32 - 4056x3040)                                                                                                                                                                                                                                                                 
   - R16 (32x32 - 4056x3040)                                                                                                                                                                                                                                                                
   - MONO_PISP_COMP1 (32x32 - 4056x3040)                                                                                                                                                                                                                                                    
   - SGRBG10 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SGBRG10 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SBGGR10 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SRGGB10 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - NV21 (32x32 - 4056x3040)                                                                                                                                                                                                                                                               
   - SBGGR8 (32x32 - 4056x3040)                                                                                                                                                                                                                                                             
   - SGRBG12 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SGBRG12 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SBGGR12 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SRGGB12 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - YUV420 (32x32 - 4056x3040)                                                                                                                                                                                                                                                             
   - NV12 (32x32 - 4056x3040)                                                                                                                                                                                                                                                               
   - YVU420 (32x32 - 4056x3040)                                                                                                                                                                                                                                                             
   - SBGGR16 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - BGGR_PISP_COMP1 (32x32 - 4056x3040)                                                                                                                                                                                                                                                    
   - SGRBG14 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SGBRG14 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SBGGR14 (32x32 - 4056x3040)                                                                                                                                                                                                                                                            
   - SRGGB14 (32x32 - 4056x3040)                                                                                                                
   - XBGR8888 (32x32 - 4056x3040)                                                                                                               
   - BGR888 (32x32 - 4056x3040)                                                                                                                 
   - RGB888 (32x32 - 4056x3040)                                                                                                                 
   - XRGB8888 (32x32 - 4056x3040)                                                                                                               
   - YUV444 (32x32 - 4056x3040)                                                                                                                 
   - YVU444 (32x32 - 4056x3040)                                                                                                                 
   - SGBRG16 (32x32 - 4056x3040)                                                                                                                
   - GBRG_PISP_COMP1 (32x32 - 4056x3040)                                                                                                        
   - SGRBG16 (32x32 - 4056x3040)                                                                                                                
   - GRBG_PISP_COMP1 (32x32 - 4056x3040)                                                                                                        
   - YUV422 (32x32 - 4056x3040)                                                                                                                 
   - YVU422 (32x32 - 4056x3040)                                                                                                                 
   - SRGGB16 (32x32 - 4056x3040)                                                                                                                
   - RGGB_PISP_COMP1 (32x32 - 4056x3040)                                                                                                        
   - BGR161616 (32x32 - 4056x3040)                                                                                                              
   - RGB161616 (32x32 - 4056x3040)                                                                                                              
   - SRGGB8 (32x32 - 4056x3040)                                                                                                                 
   - SGRBG8 (32x32 - 4056x3040)                                                                                                                 
   - SGBRG8 (32x32 - 4056x3040)                                                                                                                 
   - YVYU (32x32 - 4056x3040)                                                                                                                   
   - YUYV (32x32 - 4056x3040)                                                                                                                   
   - VYUY (32x32 - 4056x3040)                                                                                                                   
   - UYVY (32x32 - 4056x3040)                                                                                                                   
[INFO] [1773768403.204731287] [camera]:                                                                                                                                                                                                                                                     
>> stream formats:                                                     
   - R8 (32x32 - 4056x3040)                                            
   - R16 (32x32 - 4056x3040)                                           
   - MONO_PISP_COMP1 (32x32 - 4056x3040)                               
   - NV21 (32x32 - 4056x3040)                                          
   - SBGGR8 (32x32 - 4056x3040)                                        
   - SBGGR16 (32x32 - 4056x3040)                                       
   - BGGR_PISP_COMP1 (32x32 - 4056x3040)                               
   - XBGR8888 (32x32 - 4056x3040)                                      
   - BGR888 (32x32 - 4056x3040)                                        
   - RGB888 (32x32 - 4056x3040)                                        
   - XRGB8888 (32x32 - 4056x3040)                                      
   - SGBRG16 (32x32 - 4056x3040)                                       
   - GBRG_PISP_COMP1 (32x32 - 4056x3040)                               
   - SGRBG16 (32x32 - 4056x3040)                                       
   - GRBG_PISP_COMP1 (32x32 - 4056x3040)                               
   - SRGGB16 (32x32 - 4056x3040)                                       
   - RGGB_PISP_COMP1 (32x32 - 4056x3040)                               
   - BGR161616 (32x32 - 4056x3040)                                     
   - RGB161616 (32x32 - 4056x3040)                                     
   - SRGGB8 (32x32 - 4056x3040)                                        
   - SGRBG8 (32x32 - 4056x3040)                                        
   - SGBRG8 (32x32 - 4056x3040)                                        
   - YUYV (32x32 - 4056x3040)                                          
   - UYVY (32x32 - 4056x3040)                                          
[WARN] [1773768403.204767047] [camera]: no pixel format selected, auto-selecting: "XRGB8888"                                                  
[WARN] [1773768403.204780880] [camera]: set parameter 'format' to silence this warning                                                        
[INFO] [1773768403.204883936] [camera]:                                
>> XRGB8888 format sizes:                                              
   - 160x120                                                           
   - 240x160                                                           
   - 320x240                                                           
   - 400x240                                                           
   - 480x320                                                           
   - 640x360                                                           
   - 640x480                                                           
   - 720x480                                                           
   - 768x480                                                           
   - 854x480                                                           
   - 720x576                                                           
   - 800x600                                                           
   - 960x540                                                           
   - 1024x576                                                          
   - 960x640                                                           
   - 1024x600                                                          
   - 1024x768                                                          
   - 1280x720                                                          
   - 1152x864                                                          
   - 1280x800                                                          
   - 1360x768                                                          
   - 1366x768                                                          
   - 1440x900                                                          
   - 1280x1024                                                         
   - 1536x864                                                          
   - 1280x1080                                                         
   - 1600x900                                                          
   - 1400x1050                                                         
   - 1680x1050                                                         
   - 1600x1200                                                         
   - 1920x1080                                                         
   - 2048x1080                                                         
   - 1920x1200                                                         
   - 2160x1080                                                         
   - 2048x1152                                                         
   - 2560x1080                                                         
   - 2048x1536                                                         
   - 2560x1440                                                         
   - 2560x1600                                                         
   - 3840x1080                                                         
   - 2960x1440                                                         
   - 3440x1440                                                         
   - 2560x2048                                                         
   - 3200x1800                                                         
   - 3840x1600                                                         
   - 3200x2048
   - 3200x2400
   - 3840x2160
   - 3840x2400
[WARN] [1773768403.206266621] [camera]: no dimensions selected, auto-selecting: "800x600"
[WARN] [1773768403.206283880] [camera]: set parameter 'width' or 'height' to silence this warning
[0:55:12.051897857] [58]  INFO Camera camera.cpp:1215 configuring streams: (0) 800x600-XRGB8888/sRGB
[0:55:12.052095728] [72]  INFO RPI pisp.cpp:1485 Sensor: /base/axi/pcie@1000120000/rp1/i2c@88000/imx500@1a - Selected sensor format: 2028x1520-SRGGB10_1X10/RAW - Selected CFE format: 2028x1520-PC1R/RAW
[INFO] [1773768403.207339713] [camera]: camera "/base/axi/pcie@1000120000/rp1/i2c@88000/imx500@1a" configured with 800x600-XRGB8888/sRGB stream
[INFO] [1773768403.207587028] [camera]: using default calibration URL
[INFO] [1773768403.207882991] [camera]: camera calibration URL: file:///root/.ros/camera_info/imx500__base_axi_pcie_1000120000_rp1_i2c_88000_imx500_1a_800x600.yaml
[DEBUG] [1773768403.208195158] [camera]: reading camera calibration from /root/.ros/camera_info/imx500__base_axi_pcie_1000120000_rp1_i2c_88000_imx500_1a_800x600.yaml
[ERROR] [1773768403.208668528] [camera_calibration_parsers]: Unable to open camera calibration file [/root/.ros/camera_info/imx500__base_axi_pcie_1000120000_rp1_i2c_88000_imx500_1a_800x600.yaml]
[WARN] [1773768403.208941010] [camera]: Camera calibration file /root/.ros/camera_info/imx500__base_axi_pcie_1000120000_rp1_i2c_88000_imx500_1a_800x600.yaml not found
[DEBUG] [1773768403.209362343] [camera]: declare 'AeExposureMode' (type: integer)
[DEBUG] [1773768403.209792917] [camera]: declare 'Contrast' (type: double)
[DEBUG] [1773768403.210222306] [camera]: declare 'AeConstraintMode' (type: integer)
[DEBUG] [1773768403.210746232] [camera]: declare 'ExposureTimeMode' (type: integer)
[DEBUG] [1773768403.211278713] [camera]: declare 'HdrMode' (type: integer)
[DEBUG] [1773768403.211856250] [camera]: declare 'AeMeteringMode' (type: integer)
[DEBUG] [1773768403.212358436] [camera]: declare 'AeFlickerPeriod' (type: integer)
[DEBUG] [1773768403.212932695] [camera]: declare 'AnalogueGainMode' (type: integer)
[DEBUG] [1773768403.213411750] [camera]: declare 'NoiseReductionMode' (type: integer)
[DEBUG] [1773768403.213927436] [camera]: declare 'AnalogueGain' (type: double)
[DEBUG] [1773768403.214448621] [camera]: declare 'ColourCorrectionMatrix' (type: double_array)
[DEBUG] [1773768403.215024269] [camera]: declare 'Brightness' (type: double)
[DEBUG] [1773768403.215563991] [camera]: declare 'StatsOutputEnable' (type: bool)
[DEBUG] [1773768403.216060658] [camera]: declare 'ExposureTime' (type: integer)
[DEBUG] [1773768403.216378065] [camera]: declare 'CnnEnableInputTensor' (type: bool)
[DEBUG] [1773768403.216998861] [camera]: declare 'AwbEnable' (type: bool)
[DEBUG] [1773768403.217529565] [camera]: declare 'ColourGains' (type: double_array)
[DEBUG] [1773768403.218003750] [camera]: declare 'AwbMode' (type: integer)
[DEBUG] [1773768403.218544528] [camera]: declare 'ColourTemperature' (type: integer)
[DEBUG] [1773768403.219170584] [camera]: declare 'Saturation' (type: double)
[DEBUG] [1773768403.219612065] [camera]: declare 'SyncFrames' (type: integer)
[DEBUG] [1773768403.220110510] [camera]: declare 'SyncMode' (type: integer)
camera_node: /usr/local/include/libcamera/libcamera/controls.h:204: T libcamera::ControlValue::get() const [with T = libcamera::Span<const long int, 18446744073709551615>; typename std::enable_if<(libcamera::details::is_span<U>::value || std::is_same<std::basic_string_view<char>, typename std::remove_cv< <template-parameter-1-1> >::type>::value), std::nullptr_t>::type <anonymous> = nullptr]: Assertion `isArray_' failed.
                                                                                                                                                                                                                                          
```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -16,6 +16,7 @@ set(PACKAGE_DEPENDENCIES
   "rclcpp"
   "rclcpp_components"
   "sensor_msgs"
+  "diagnostic_msgs"
   "camera_info_manager"
   "cv_bridge"
 )
@@ -77,6 +78,7 @@ target_link_libraries(camera_component PUBLIC
   rclcpp::rclcpp
   rclcpp_components::component
   ${sensor_msgs_TARGETS}
+  ${diagnostic_msgs_TARGETS}
   camera_info_manager::camera_info_manager
   cv_bridge::cv_bridge
 )

--- a/package.xml
+++ b/package.xml
@@ -13,6 +13,7 @@
   <depend>rclcpp</depend>
   <depend>rclcpp_components</depend>
   <depend>sensor_msgs</depend>
+  <depend>diagnostic_msgs</depend>
   <depend>camera_info_manager</depend>
   <depend>cv_bridge</depend>
 

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -642,6 +642,25 @@ CameraNode::process(libcamera::Request *const request)
     if (!running)
       return;
 
+    // prepare message header
+    std_msgs::msg::Header hdr;
+    hdr.frame_id = frame_id;
+
+    // if using sensor timestamps, get the sensor timestamp from the request metadata
+    int64_t sensor_latency = 0;
+    if (!use_node_time) {
+      const libcamera::ControlList &req_metadata = request->metadata();
+      if (const std::optional<int64_t> sensor_ts = req_metadata.get(libcamera::controls::SensorTimestamp)) {
+        sensor_latency = rclcpp::Clock(RCL_STEADY_TIME).now().nanoseconds() - sensor_ts.value();
+      }
+      else {
+        RCLCPP_WARN_STREAM_ONCE(get_logger(), "sensor timestamp not available, falling back to node time as reference");
+      }
+    }
+
+    // Adjust timestamp by the sensor latency
+    hdr.stamp = this->now() - rclcpp::Duration::from_nanoseconds(sensor_latency);
+
     if (request->status() == libcamera::Request::RequestComplete) {
       assert(request->buffers().size() == 1);
 
@@ -651,25 +670,6 @@ CameraNode::process(libcamera::Request *const request)
       size_t bytesused = 0;
       for (const libcamera::FrameMetadata::Plane &plane : metadata.planes())
         bytesused += plane.bytesused;
-
-      // prepare message header
-      std_msgs::msg::Header hdr;
-      hdr.frame_id = frame_id;
-
-      // if using sensor timestamps, get the sensor timestamp from the request metadata
-      int64_t sensor_latency = 0;
-      if (!use_node_time) {
-        const libcamera::ControlList &req_metadata = request->metadata();
-        if (const std::optional<int64_t> sensor_ts = req_metadata.get(libcamera::controls::SensorTimestamp)) {
-          sensor_latency = rclcpp::Clock(RCL_STEADY_TIME).now().nanoseconds() - sensor_ts.value();
-        }
-        else {
-          RCLCPP_WARN_STREAM_ONCE(get_logger(), "sensor timestamp not available, falling back to node time as reference");
-        }
-      }
-
-      // Adjust timestamp by the sensor latency
-      hdr.stamp = this->now() - rclcpp::Duration::from_nanoseconds(sensor_latency);
 
       // prepare image messages
       const libcamera::StreamConfiguration &cfg = stream->configuration();

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -15,6 +15,7 @@
 #include <cv_bridge/cv_bridge.h>
 #endif
 #include <atomic>
+#include <diagnostic_msgs/msg/diagnostic_array.hpp>
 #include <iostream>
 #include <libcamera/base/shared_fd.h>
 #include <libcamera/base/signal.h>
@@ -100,6 +101,7 @@ private:
   rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_image;
   rclcpp::Publisher<sensor_msgs::msg::CompressedImage>::SharedPtr pub_image_compressed;
   rclcpp::Publisher<sensor_msgs::msg::CameraInfo>::SharedPtr pub_ci;
+  rclcpp::Publisher<diagnostic_msgs::msg::DiagnosticArray>::SharedPtr pub_diagnostics;
 
   camera_info_manager::CameraInfoManager cim;
 
@@ -329,6 +331,8 @@ CameraNode::CameraNode(const rclcpp::NodeOptions &options)
   pub_image_compressed =
     this->create_publisher<sensor_msgs::msg::CompressedImage>("~/image_raw/compressed", 1);
   pub_ci = this->create_publisher<sensor_msgs::msg::CameraInfo>("~/camera_info", 1);
+  pub_diagnostics =
+    this->create_publisher<diagnostic_msgs::msg::DiagnosticArray>("/diagnostics", 1);
 
   // start camera manager and check for cameras
   const int ec_start = camera_manager.start();
@@ -661,8 +665,23 @@ CameraNode::process(libcamera::Request *const request)
     // Adjust timestamp by the sensor latency
     hdr.stamp = this->now() - rclcpp::Duration::from_nanoseconds(sensor_latency);
 
+    diagnostic_msgs::msg::DiagnosticArray diagnostic_array;
+    diagnostic_array.header = hdr;
+
+    diagnostic_msgs::msg::DiagnosticStatus diagnostic_status;
+    diagnostic_status.hardware_id = camera->id();
+
     if (request->status() == libcamera::Request::RequestComplete) {
       assert(request->buffers().size() == 1);
+
+      diagnostic_status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
+
+      for (const auto &[id, value] : request->metadata()) {
+        diagnostic_msgs::msg::KeyValue kv;
+        kv.key = libcamera::controls::controls.at(id)->name();
+        kv.value = value.toString();
+        diagnostic_status.values.push_back(kv);
+      }
 
       // get the stream and buffer from the request
       const libcamera::FrameBuffer *buffer = request->findBuffer(stream);
@@ -725,8 +744,13 @@ CameraNode::process(libcamera::Request *const request)
       pub_ci->publish(ci);
     }
     else if (request->status() == libcamera::Request::RequestCancelled) {
+      diagnostic_status.level = diagnostic_msgs::msg::DiagnosticStatus::WARN;
       RCLCPP_ERROR_STREAM(get_logger(), "request '" << request->toString() << "' cancelled");
     }
+
+    diagnostic_array.status.push_back(diagnostic_status);
+
+    pub_diagnostics->publish(diagnostic_array);
 
     // redeclare implicitly undeclared parameters
     parameter_handler.redeclare();

--- a/src/CameraNode.cpp
+++ b/src/CameraNode.cpp
@@ -676,7 +676,11 @@ CameraNode::process(libcamera::Request *const request)
 
       diagnostic_status.level = diagnostic_msgs::msg::DiagnosticStatus::OK;
 
-      for (const auto &[id, value] : request->metadata()) {
+      // publish metadata, sorted by control ID
+      for (const auto &[id, value] : std::map {
+             request->metadata().begin(),
+             request->metadata().end(),
+           }) {
         diagnostic_msgs::msg::KeyValue kv;
         kv.key = libcamera::controls::controls.at(id)->name();
         kv.value = value.toString();

